### DISCUSSION TITLE
Make this chatbot more flexible

### DISCUSSION THREAD

**Discussion by beralc** [mayooear/ai-pdf-chatbot-langchain#144] [Q&A]
Hi all,
I've managed to use this app, but since I don't have access to GPT4 i'm using GPT3.5 turbo. The answers are always "Hmm, I'm not sure. The context provided does not give a clear answer to the question  ..." and then it shows sources. So what could the problem be? And.. is there a way to also use chatgpt if the information isn't in the embeddings?

**Comment by fmcreativeco** [comment]
You need to edit [makechain.ts](https://github.com/mayooear/gpt4-pdf-chatbot-langchain/blob/main/utils/makechain.ts) and modify the default prompt to fit your use case. Every question is a new request to the GPT3.5 Turbo API, and each request to the API has no memory of your previous request. So your prompt needs to include detailed instructions of how the API should interpret _each and every question_ you send to it, from scratch. 

The code in this repo sends some of the details of your most recent question/answer back to the API with your next question, which can lead you to the "Hmm, I'm not sure" problem due to the way the default prompt is written. If you ask it about color options of a Ford Mustang, and then follow up with a question about the price of Hondas, the "context" has been limited to the previous discussion about Ford Mustangs... so it's now going to struggle finding an answer about Hondas.  

You can fix this by modifying the default prompt to better handle your use case. For example you can remove the instructions about responding with "Hmm, I'm not sure" if the answer can't be found in the _recent_ context, and instead request that if the answer can't be found in the recent context, to disregard recent context and look for the answer using the entire document. This approach should fix the Mustang and Honda example I mentioned above. But prompt engineering is an art, not a science, and will require a lot of experimentation to get it to work for your use case. 

Hope this helps, good luck!

**Reply by K1ngSk1n** [reply]
You can also increase the temperature (between 0(most factual) and 1(most imaginative) to encourage GPT to be more interpretive with it's responses. Same again in makechain.ts. I am currently using 0.5

**Comment by groovybits** [comment]
My fork is very flexible, you can alter the prompts inline :D there's a lot of setup and not as user friendly if you want to not do the setup unfortunately. I am going to work on disabling all the bells and whistles by default soon. It basically extends the chat modes to all the use cases one would mostly have. The prompts are there to lift out and ways to handle history for different modes. Has the temperature set to do what I can tell are the best behaviors on question/answer off facts vs. story telling and making it less exact and factual / less prone to not answering and seeing it as out of context. At least it allows you to get a better idea of what can be done for prompting quickly, although also note how they are modular / stack and have common strings too.

https://github.com/groovybits/gaib/blob/main/config/personalityPrompts.ts

https://github.com/groovybits/gaib

**Reply by groovybits** [reply]
<img width="557" alt="Screenshot 2023-07-17 at 13 47 43" src="https://github.com/mayooear/gpt4-pdf-chatbot-langchain/assets/1077255/2de575ff-375c-43e7-a0dc-ac88d86efd00">
i just updated the code to no longer require all the different services, no need for auth through firebase/firestore anymore unless setup in env variables.

I am adding chroma soon, and the default chatbot mode (boring but fine :D). I plan on really making it extensible more and more, enterprise ready with full GCP Cloudrun support and all the other bells and whistles added. Yet also be able to run stand alone just in a chatbot like chatGPT and this chatbot output. GAIB has Themes, and the main theme is more about speaking with a real sensory entity rather than a chatroom feel.

The prompting is very important for responses to fit what the use case is. So you can tinker with GAIB like a prompt scope and change them while running easily in the HTML forms, iterate and discover what fits perfectly.

**Comment by alejo-perez-upc-77** [comment]
You can:

- Increase the model's temperature as suggested above
- Modify the prompt so as it is more flexible
- Use a custom LLM Router. It would read the question (+ history optionally) and output a control variable. You would use this variable to call either the ConversationalReatrievalQAChain or another chain/LLM with a different prompt
